be/dtm0_log.c

```diff
@@ -0,0 +1,658 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+
+#define M0_TRACE_SUBSYSTEM M0_TRACE_SUBSYS_DTM0
+#include "be/dtm0_log.h"
+#include "be/alloc.h"
+#include "dtm0/clk_src.h"
+#include "lib/buf.h"
+#include "be/list.h"
+#include "be/op.h"
+#include "be/tx.h"
+#include "be/tx_credit.h"
+#include "be/seg.h"
+#include "lib/assert.h" /* M0_PRE */
+#include "lib/errno.h"  /* ENOENT */
+#include "lib/memory.h" /* M0_ALLOC */
+#include "lib/trace.h"
+#include "dtm0/fop.h"  /* dtm0_req_fop */
+#include "motr/magic.h"
+
+
+M0_TL_DESCR_DEFINE(lrec, "DTM0 Log", static, struct m0_dtm0_log_rec,
+                   u.dlr_tlink, dlr_magic, M0_BE_DTM0_LOG_REC_MAGIX,
+                   M0_BE_DTM0_LOG_MAGIX);
+M0_TL_DEFINE(lrec, static, struct m0_dtm0_log_rec);
+
+
+M0_BE_LIST_DESCR_DEFINE(lrec, "DTM0 PLog", static, struct m0_dtm0_log_rec,
+			u.dlr_link, dlr_magic, M0_BE_DTM0_LOG_REC_MAGIX,
+			M0_BE_DTM0_LOG_MAGIX);
+M0_BE_LIST_DEFINE(lrec, static, struct m0_dtm0_log_rec);
+
+
+static bool m0_be_dtm0_log__invariant(const struct m0_be_dtm0_log *log)
+{
+	return _0C(log != NULL) &&
+	       _0C(log->dl_cs != NULL); // &&
+	/* TODO: Add an invariant check against the volatile part */
+	       /* _0C(lrec_tlist_invariant(log->u.dl_tlist)); */
+}
+
+static bool m0_dtm0_log_rec__invariant(const struct m0_dtm0_log_rec *rec)
+{
+	return _0C(rec != NULL) &&
+	       _0C(m0_dtm0_tx_desc__invariant(&rec->dlr_txd)); // &&
+	       /* _0C(m0_tlink_invariant(&lrec_tl, rec)); */
+}
+
+/*
+ * Allocate memory for a dtm0 volatile log structure.
+ */
+M0_INTERNAL int m0_be_dtm0_log_alloc(struct m0_be_dtm0_log **out)
+{
+	struct m0_be_dtm0_log *log;
+
+	M0_PRE(out != NULL);
+
+	M0_ALLOC_PTR(log);
+	if (log == NULL)
+		return M0_ERR(-ENOMEM);
+
+	M0_ALLOC_PTR(log->u.dl_tlist);
+	if (log->u.dl_tlist == NULL) {
+		m0_free(log);
+		return M0_ERR(-ENOMEM);
+	}
+	lrec_tlist_init(log->u.dl_tlist);
+	*out = log;
+	return 0;
+}
+
+/* Intialize a dtm0 log. *out should contain a pointer to a pre-allocated
+ * dtm0_log structure. */
+M0_INTERNAL int m0_be_dtm0_log_init(struct m0_be_dtm0_log  *log,
+                                    struct m0_dtm0_clk_src *cs,
+                                    bool                    is_plog)
+{
+	M0_PRE(log != NULL);
+	M0_PRE(cs != NULL);
+
+	m0_mutex_init(&log->dl_lock);
+	log->dl_is_persistent = is_plog;
+	log->dl_cs = cs;
+	return 0;
+}
+
+M0_INTERNAL void m0_be_dtm0_log_fini(struct m0_be_dtm0_log **in_log)
+{
+	struct m0_be_dtm0_log *log = *in_log;
+
+	M0_PRE(m0_be_dtm0_log__invariant(log));
+	m0_mutex_fini(&log->dl_lock);
+	lrec_tlist_fini(log->u.dl_tlist);
+	log->dl_cs = NULL;
+}
+
+M0_INTERNAL void m0_be_dtm0_log_free(struct m0_be_dtm0_log **in_log)
+{
+	struct m0_be_dtm0_log *log = *in_log;
+	bool                   ispstore = log->dl_is_persistent;
```

Comment:
No need to define a variable that is only used in a single place.

Reply:
Done.